doc/read-the-docs-site/adr/0005-inlining-saturated-functions.md

```diff
@@ -0,0 +1,185 @@
+in f 5
+```
+
+With beta-reduction, `f` becomes `\y.4+y` and it has 1 lambda.
+The _body_ `f 5` is a fully applied function!
+We can reduce it to 9.
+However, because the _rhs_ `(\x.\y.x+y) 4` is a function application, not a lambda abstraction,
+we won't inline it with the current implementation. This is something we may be able to improve.
+
+### The implementation: Counting lambdas and applications
+
+We run `countLam` to the _rhs_ of the binding:
+
+```haskell
+-- | Count the number of type and term lambdas in the RHS of a binding
+countLam :: 
+    Term tyname name uni fun a -- ^ the RHS
+    -> (Natural, Natural) -- ^ Number of type lambdas, number of term lambdas
+countLam = countLamInner 0 0
+    where
+      countLamInner ::
+        Natural -- ^ Number of type lambdas of the function.
+        -> Natural -- ^ Number of term lambdas of the function.
+        -> Term tyname name uni fun a -- ^ The rhs term that is being counted.
+        -> (Natural, Natural)
+      countLamInner typeLambdas termLambdas (LamAbs _a _n _ty body) =
+        -- If the term is a term lambda abstraction, increase the count of the number of term lambdas by one.
+        -- Keep on examining the body.
+        countLamInner typeLambdas (termLambdas + 1) body
+      countLamInner typeLambdas termLambdas (TyAbs _a _n _kind body) =
+        -- If the term is a type lambda abstraction, increase the count of the number of type lambdas by one.
+        -- Keep on examining the body.
+        countLamInner (typeLambdas + 1) termLambdas body
+      countLamInner typeLambdas termLambdas _ = 
+        -- whenever we encounter a body that is not a lambda abstraction, we are done counting
+        (typeLambdas, termLambdas) 
+```
+
+To count the number of applied arguments, we already have something that collects the applications
+([extractApps](
+https://github.com/input-output-hk/plutus/blob/791e1dc0f6f5b9954da4c4d19e4597f5e75d0727/plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/Inline.hs#L138))
+in the UPLC inliner. We can run `length` to the result to get the number for the UPLC inliner.
+It's straight forward to implement this for the PIR inliner as well.
+
+### The implementation: call site inlining
+
+Another substantial component of this implementation is the actual inlining at the call site.
+Details of this to be worked out.
+
+## Future plans
+
+We have plans to improve this pass so that it will catch more cases that are desirable,
+and not inline undesirably.
+
+### Add beta-reduction pass
+
+If beta-reduction of _rhs_ is run before this pass, it can catch cases categorized by example 3.
+
+### Add inline size heuristic
+
+If we do some analysis of the inline size, we can better decide whether it's desirable to inline.
+Plutonomy also [uses a similar heuristic](https://github.com/well-typed/plutonomy/blob/14b9bd46084db1b785b3a99d55f7f10d38165ee8/src/Plutonomy/Hereditary/Transform.hs#L266).
+
+### Add more refined heuristic for call site inlining
```

Comment:
I think we should just delete it, inlining of saturated functions is a superior version anyway.